File: solana/src/main/java/org/p2p/solanaj/core/Transaction.java
package org.p2p.solanaj.core;

import org.bitcoinj.core.Base58;
import org.p2p.solanaj.utils.ShortvecEncoding;
import org.p2p.solanaj.utils.TweetNaclFast;

import java.math.BigInteger;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Transaction {

    public static final int SIGNATURE_LENGTH = 64;
    public static final byte[] DEFAULT_SIGNATURE = new byte[64];

    private final Message message;
    private final List<Signature> signatures;
    private byte[] serializedMessage;
    private PublicKey feePayer;

    public Transaction() {
        this.message = new Message();
        this.signatures = new ArrayList<>();
    }

    public Transaction addInstruction(TransactionInstruction instruction) {
        message.addInstruction(instruction);
        return this;
    }

    public Transaction addInstructions(List<TransactionInstruction> instructions) {
        for (TransactionInstruction instruction : instructions) {
            message.addInstruction(instruction);
        }
        return this;
    }

    public List<TransactionInstruction> getInstructions() {
        return message.getInstructions();
    }

    public void setRecentBlockHash(String recentBlockhash) {
        message.setRecentBlockHash(recentBlockhash);
    }

    public String getRecentBlockHash() {
        return message.getRecentBlockHash();
    }

    public void setFeePayer(PublicKey feePayer) {
        message.setFeePayer(feePayer);
        this.feePayer = feePayer;
    }

    public PublicKey getFeePayer() {
        return feePayer;
    }

    public void sign(Account signer) {
        sign(Arrays.asList(signer));
    }

    public void signWithoutSignatures(PublicKey owner) {

        if (feePayer == null) {
            feePayer = owner;
        }
        message.setFeePayer(feePayer);

        serializedMessage = message.serialize();
    }

    public void sign(List<Account> signers) {

        if (signers.size() == 0) {
            throw new IllegalArgumentException("No signers");
        }

        if (feePayer == null) {
            feePayer = signers.get(0).getPublicKey();
        }
        message.setFeePayer(feePayer);

        serializedMessage = message.serialize();

        for (Account signer : signers) {
            TweetNaclFast.Signature signatureProvider = new TweetNaclFast.Signature(new byte[0], signer.getKeypair());
            byte[] signature = signatureProvider.detached(serializedMessage);

            Signature newSignature = new Signature(signer.getPublicKey(), Base58.encode(signature));
            signatures.add(newSignature);
        }
    }

    public List<AccountMeta> getAccountKeys() {
        return message.getAccountKeys();
    }

    public byte[] serialize() {
        int signaturesSize = signatures.size();
        byte[] signaturesLength = ShortvecEncoding.encodeLength(signaturesSize);

        ByteBuffer out = ByteBuffer
                .allocate(signaturesLength.length + signaturesSize * SIGNATURE_LENGTH + serializedMessage.length);

        out.put(signaturesLength);

        for (Signature signature : signatures) {
            if (signature.getSignature() != null) {
                byte[] rawSignature = Base58.decode(signature.getSignature());
                out.put(rawSignature);
            }
        }

        out.put(serializedMessage);

        return out.array();
    }

    public Signature getSignature() {
        if (signatures.size() > 0) {
            return signatures.get(0);
        }

        return null;
    }

    public Signature findSignature(PublicKey publicKey) {
        for (Signature signature : signatures) {
            if (signature.getPublicKey().equals(publicKey)) {
                return signature;
            }
        }

        return null;
    }

    public List<Signature> getAllSignatures() {
        return signatures;
    }

    public BigInteger calculateTransactionFee(BigInteger lamportsPerSignatures) {
        message.serialize();
        return BigInteger.valueOf((long) message.getNumRequiredSignatures()).multiply(lamportsPerSignatures);
    }
}

File: solana/src/main/java/org/p2p/solanaj/core/Transaction.kt
package org.p2p.solanaj.core

import org.bitcoinj.core.Base58
import java.math.BigInteger
import java.nio.ByteBuffer
import org.p2p.solanaj.utils.ShortvecEncoding
import org.p2p.solanaj.utils.TweetNaclFast

private const val SIGNATURE_LENGTH = 64

class Transaction {

    private val message: Message = Message()
    private val signatures = mutableListOf<Signature>()
    private var feePayer: PublicKey? = null

    private lateinit var serializedMessage: ByteArray

    val accountKeys: List<AccountMeta>
        get() = message.accountKeys

    val instructions: List<TransactionInstruction>
        get() = message.instructions

    val recentBlockHash: String
        get() = message.recentBlockHash

    val signature: Signature?
        get() = signatures.firstOrNull()

    val allSignatures: List<Signature>
        get() = signatures

    fun addInstruction(instruction: TransactionInstruction?): Transaction {
        message.addInstruction(instruction)
        return this
    }

    fun addInstructions(instructions: List<TransactionInstruction?>): Transaction {
        for (instruction in instructions) {
            message.addInstruction(instruction)
        }
        return this
    }

    fun setRecentBlockhash(recentBlockhash: String) {
        message.recentBlockHash = recentBlockhash
    }

    fun setFeePayer(feePayer: PublicKey) {
        message.setFeePayer(feePayer)
        this.feePayer = feePayer
    }

    fun getFeePayer(): PublicKey? = feePayer

    fun sign(signer: Account) {
        sign(listOf(signer))
    }

    fun signWithoutSignatures(owner: PublicKey) {
        if (feePayer == null) {
            feePayer = owner
        }
        message.setFeePayer(feePayer)
        serializedMessage = message.serialize()
    }

    fun sign(signers: List<Account>) {
        require(signers.isNotEmpty()) { "No signers" }
        if (feePayer == null) {
            feePayer = signers[0].publicKey
        }
        message.setFeePayer(feePayer)
        serializedMessage = message.serialize()
        for (signer in signers) {
            val signatureProvider = TweetNaclFast.Signature(ByteArray(0), signer.keypair)
            val signature = signatureProvider.detached(serializedMessage)
            val newSignature = Signature(signer.publicKey, Base58.encode(signature))
            signatures.add(newSignature)
        }
    }

    fun serialize(): ByteArray {
        val signaturesSize = signatures.size
        val signaturesLength = ShortvecEncoding.encodeLength(signaturesSize)
        val capacity = signaturesLength.size + signaturesSize * SIGNATURE_LENGTH + serializedMessage.size
        val out = ByteBuffer.allocate(capacity)
        out.put(signaturesLength)

        signatures.forEach { signature ->
            val rawSignature = Base58.decode(signature.signature)
            out.put(rawSignature)
        }

        out.put(serializedMessage)
        return out.array()
    }

    fun findSignature(publicKey: PublicKey): Signature? =
        signatures.find { it.publicKey.equals(publicKey) }

    fun addSignature(signature: Signature) {
        val index = signatures.indexOfFirst { it.publicKey.equals(signature.publicKey) }

        // FIXME: Perhaps we need to throw exception here because there feePayer should exist
        if (index == -1) {
            signatures.add(0, signature)
            return
        }

        signatures[index] = signature
    }

    fun calculateTransactionFee(lamportsPerSignatures: BigInteger): BigInteger {
        message.serialize()
        return BigInteger.valueOf(message.numRequiredSignatures.toLong()).multiply(lamportsPerSignatures)
    }
}
